File: android/src/main/kotlin/com/example/pl_image_downloader/models/DownloadTask.kt
package com.example.pl_image_downloader.models

import com.example.pl_image_downloader.models.enum.DownloadException


/**
 *  * Data class representing a download task with its associated properties.
 */
data class DownloadTask(
    val id: Long? = null,
    val enqueueId: Long? = null,
    val url: String,
    val destinationPath: String = "",
    val fileName: String,
    val overwrite: Boolean = false,
    val downloadStatus: DownloadStatus = DownloadStatus.IDLE,
    val progress: Int = 0,

    /** Handle exception if any occurred during the download process. */
    val exception: DownloadException? = null,

    /** Result of the download operation. */
    val result: DownloadResult? = null
) {
    companion object {
        /** * Extension function to convert DownloadInfo to DownloadTask. */
        fun fromDownloadInfo(info: DownloadInfo): DownloadTask {
            return DownloadTask(id = info.id, url = info.url, fileName = info.fileName)
        }
    }

    /** * Marks the download task as successful and updates its properties accordingly. */
    fun success(): DownloadTask {
        return this.copy(
            downloadStatus = DownloadStatus.COMPLETED,
            progress = 100,
            result = DownloadResult(
                path = destinationPath,
                dictionary = destinationPath.substringBeforeLast("/"),
                fileName = fileName,
                isSuccess = true
            )
        )
    }

    /** * Marks the download task as failed and updates its properties accordingly. */
    fun failed(errorMessage: String? = null, exception: DownloadException): DownloadTask {
        return this.copy(
            downloadStatus = DownloadStatus.FAILED,
            exception = exception,
            result = DownloadResult(
                path = destinationPath,
                dictionary = destinationPath.substringBeforeLast("/"),
                fileName = fileName,
                isSuccess = false,
                errorMessage = errorMessage ?: exception.name
            )
        )
    }
}

File: ios/Classes/Models/DownloadTask.swift
//
//  DownloadTask.swift
//

import Foundation

struct DownloadTask {

    let id: Int64?
    let enqueueId: Int64?
    let url: String
    let destinationPath: String
    let fileName: String
    let overwrite: Bool
    let downloadStatus: DownloadStatus
    let progress: Int

    /// Result of the download operation.
    let result: DownloadResult?

    init(
        id: Int64? = nil,
        enqueueId: Int64? = nil,
        url: String,
        destinationPath: String = "",
        fileName: String,
        overwrite: Bool = false,
        downloadStatus: DownloadStatus = .idle,
        progress: Int = 0,
        result: DownloadResult? = nil
    ) {
        self.id = id
        self.enqueueId = enqueueId
        self.url = url
        self.destinationPath = destinationPath
        self.fileName = fileName
        self.overwrite = overwrite
        self.downloadStatus = downloadStatus
        self.progress = progress
        self.result = result
    }
}

extension DownloadTask {

    static func fromDownloadInfo(_ info: DownloadInfo) -> DownloadTask {

        return DownloadTask(
            id: info.id,
            url: info.url,
            fileName: info.fileName
        )
    }

    func success() -> DownloadTask {

        let directory = URL(fileURLWithPath: destinationPath)
            .deletingLastPathComponent()
            .path

        return DownloadTask(
            id: id,
            enqueueId: enqueueId,
            url: url,
            destinationPath: destinationPath,
            fileName: fileName,
            overwrite: overwrite,
            downloadStatus: .completed,
            progress: 100,
            result: DownloadResult(
                path: destinationPath,
                dictionary: directory,
                fileName: fileName,
                isSuccess: true,
                error: nil
            )
        )
    }

    func failed(errorMessage: String? = nil) -> DownloadTask {

        let directory = URL(fileURLWithPath: destinationPath)
            .deletingLastPathComponent()
            .path

        return DownloadTask(
            id: id,
            enqueueId: enqueueId,
            url: url,
            destinationPath: destinationPath,
            fileName: fileName,
            overwrite: overwrite,
            downloadStatus: .failed,
            progress: progress,
            result: DownloadResult(
                path: destinationPath,
                dictionary: directory,
                fileName: fileName,
                isSuccess: false,
                error: errorMessage ?? "Unknown error occurred during download."
            )
        )
    }
    
    func copy(
        id: Int64?? = nil,
        enqueueId: Int64?? = nil,
        url: String? = nil,
        destinationPath: String? = nil,
        fileName: String? = nil,
        overwrite: Bool? = nil,
        downloadStatus: DownloadStatus? = nil,
        progress: Int? = nil,
        result: DownloadResult?? = nil
    ) -> DownloadTask {
        return DownloadTask(
            id: id ?? self.id,
            enqueueId: enqueueId ?? self.enqueueId,
            url: url ?? self.url,
            destinationPath: destinationPath ?? self.destinationPath,
            fileName: fileName ?? self.fileName,
            overwrite: overwrite ?? self.overwrite,
            downloadStatus: downloadStatus ?? self.downloadStatus,
            progress: progress ?? self.progress,
            result: result ?? self.result
        )
    }
}
